### ISSUE TITLE
bring tidal-listener option back

### ISSUE THREAD

**Issue by polymorphicengine** [tidalcycles/pulsar-tidalcycles#200]
hey! I just updated the tidal listener (https://github.com/tidalcycles/Tidal/pull/996) and it is now much more like ghci, i think it would be great to be able to choose it as an interpreter (hopefully in the future it will be runnable without an installed haskell environment).

currently you can send it haskell statements to evaluate, ask for types and load haskell files containing definitions. to do this you send it messages like `/eval <code>` or `/type <code>`.

what i am wondering is if i should make it even more ghci like and just have it receiv one type of message that is either an expression to evaluate or starts with `:t` or `:load`. 

on the other hand this could also be handled on the editor side of things, which do you think would be better?

**Comment by ndr-brt** [comment]
Hey, that's great to hear!
In fact tidal-listener is still available, but it would need to be updated to the new behavior.
The problem is that currently we're not able to release a new plugin version because of this: https://github.com/pulsar-edit/package-backend/issues/63

But in general can't wait for this!

**Comment by rlafuente** [comment]
I was trying to get the tidal-listener mode to work (i need flashing numbers!), but after updating everything and reinstalling tidal + tidal-listener, I get this when evaluating a block in Pulsar:

`Unhandled message: Message {messageAddress = "/code", messageDatum = [AsciiString {d_ascii_string = "d0"},AsciiString {d_ascii_string = "jux rev $\nn (scale \"minor\" (\n  struct \"t([9|13]*16,16,[1|3])\"\n  $ rot \"<[5|4|3] 2 [1|2|3] 0>\"\n  $ \"[0 2 3 4 [6|9] 7 6 [8|10]]*2\")\n  |+ \"[-12|-12|0|12]*16\"\n  |+ \"<3 -2 1 1 0 0 0 0>\"\n  )\n# sound \"superpiano\"\n# velocity \"[0.5|0.55]*16\"\n# sustain \"[4|7]*16\"\nhush"}]}`

Is this expected? I remember asking @ndr-brt about this on Discord and hearing tidal-listener should be working, but I'm a bit lost after reading this issue and the comments :)

**Comment by polymorphicengine** [comment]
it will take some time to reintegrate the current updated version of the listener with the editor plugins since the API slightly changed, it shouldn't be much work though since the listener is now quite ghci-like

**Comment by rlafuente** [comment]
Thank you for such a quick reply -- I was hoping I could use this on a showcase tomorrow, and it's useful to know I shouldn't try further for now :) and thank you for being on top of this

**Comment by ndr-brt** [comment]
hey @polymorphicengine is the new api described somewhere? I looked at the [tidal-listener readme](https://github.com/tidalcycles/Tidal/tree/077b65f482a76cd5bacfc7a21ec2ae7a0f86f0c2/tidal-listener) but it seems not changed

**Comment by polymorphicengine** [comment]
yes, it should be exactly how it is described in the link you posted. `/code` messages are deprecated and replaced with `/eval`

**Comment by rlafuente** [comment]
Ok I really really want this feature to happen and want to help make it real. 

I have some experience building pulsar plugins, but need some hand-holding on what needs changing. I expect it's more than replacing `/code` calls with `/eval`?

**Comment by ndr-brt** [comment]
> Ok I really really want this feature to happen and want to help make it real.
> 
> I have some experience building pulsar plugins, but need some hand-holding on what needs changing. I expect it's more than replacing `/code` calls with `/eval`?

hey! all the logic for handling tidal-listener should be located in the [tidal-listener-repl.js](https://github.com/tidalcycles/pulsar-tidalcycles/blob/207bb39e4c7e1789582cf35550002befee10b83b/lib/tidal-listener-repl.js) file, IIRC there still was some duplication with the `repl.js` (the standard one), but it worked with the old tidal-listener version.
Last time I tried I wasn't able to make `tidal-listener` work but I guess it was something on my side

**Comment by yaxu** [comment]
Hey just to add that I'm not sure if anyone has got highlighting working via tidal-listener yet! The mechanics are there but it might need a bit more work.
